File: legacy-go/internal/skillscmd/eval.go
package skillscmd

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/buildbetter/skillrank/internal/command"
	reg "github.com/buildbetter/skillrank/internal/registry"
	"github.com/buildbetter/skillrank/internal/registry/runner"
)

func runEval(args []string, ctx command.Context) int {
	flags := command.ParseFlags(args)
	if len(flags.Positionals) == 0 {
		fmt.Fprintln(ctx.Stderr, "usage: eval <ref> --suite <id> [--trials N] [--agent claude|codex] [--model M] [--publish]")
		return 2
	}
	ref := flags.Positionals[0]
	suiteID := flags.Values["suite"]
	if strings.TrimSpace(suiteID) == "" {
		fmt.Fprintln(ctx.Stderr, "error: --suite <id> is required")
		return 2
	}
	trials := 3
	if v := flags.Values["trials"]; v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			trials = n
		}
	}
	provider := strings.TrimSpace(flags.Values["agent"])
	if provider == "" {
		provider = detectAgentProvider()
	}
	if provider != "claude" && provider != "codex" {
		fmt.Fprintln(ctx.Stderr, "error: could not find a supported agent CLI; install `claude` or `codex`, or pass --agent")
		return 1
	}
	if _, err := exec.LookPath(provider); err != nil {
		fmt.Fprintf(ctx.Stderr, "error: agent %q not found on PATH\n", provider)
		return 1
	}

	client, err := newClient(flags)
	if err != nil {
		fmt.Fprintf(ctx.Stderr, "error: %s\n", err)
		return 1
	}

	// Resolve the skill and ensure we have its content for the treatment arm.
	resolved, err := client.Resolve(ref)
	if err != nil {
		fmt.Fprintf(ctx.Stderr, "error: %s\n", err)
		return 1
	}
	if strings.TrimSpace(resolved.InlineContent) == "" && strings.TrimSpace(resolved.RawContentURL) != "" {
		content, fetchErr := client.FetchRawContent(resolved.RawContentURL)
		if fetchErr != nil {
			fmt.Fprintf(ctx.Stderr, "error: %s\n", fetchErr)
			return 1
		}
		resolved.InlineContent = content
	}
	if strings.TrimSpace(resolved.InlineContent) == "" {
		fmt.Fprintln(ctx.Stderr, "error: registry did not provide skill content to evaluate")
		return 1
	}

	suite, err := client.GetSuite(suiteID)
	if err != nil {
		fmt.Fprintf(ctx.Stderr, "error: %s\n", err)
		return 1
	}
	verifiers, err := client.FetchVerifiers(suiteID, suite.Version)
	if err != nil {
		fmt.Fprintf(ctx.Stderr, "error: could not fetch verifiers for suite %s: %s\n", suiteID, err)
		return 1
	}

	cfg := runner.Config{Trials: trials, Model: flags.Values["model"]}

	// Cost estimate + confirmation.
	estTokens, estCost := runner.EstimateCost(suite, cfg)
	if !flags.WantsStructuredOutput() {
		fmt.Fprintf(ctx.Stdout, "Eval plan: skill %s vs no-skill on suite %s@%s\n", resolved.Slug, suite.ID, suite.Version)
		fmt.Fprintf(ctx.Stdout, "  agent: %s | model: %s | %d trials/arm | %d tasks × 2 arms\n",
			provider, orDash(cfg.Model), trials, len(suite.Tasks))
		fmt.Fprintf(ctx.Stdout, "  estimated: ~%s tokens, ~$%.2f on YOUR agent subscription\n", humanInt(estTokens), estCost)
		if !runner.DockerAvailable() {
			fmt.Fprintln(ctx.Stdout, "  note: Docker not detected → worktree isolation; results publish as Self-reported.")
		}
		if !flags.BoolValues["yes"] && !flags.BoolValues["y"] {
			if !confirm(ctx, "Proceed?") {
				fmt.Fprintln(ctx.Stdout, "Aborted.")
				return 1
			}
		}
	}

	agentRunner := &runner.CLIAgentRunner{Provider: provider, Binary: provider, Version: detectAgentVersion(provider)}
	fixtures := runner.NewGitFixtureProvider(suite.Fixture.GitURL, suite.Fixture.Commit)
	defer fixtures.Close()
	verifier := &runner.ScriptVerifier{Commands: verifiers}

	runCtx := context.Background()
	result, err := runner.RunEval(runCtx, suite, resolved, cfg, agentRunner, fixtures, verifier)
	if err != nil {
		fmt.Fprintf(ctx.Stderr, "error: %s\n", err)
		return 1
	}
	result.Bundle.CreatedAt = time.Now().UTC().Format(time.RFC3339)

	// Persist the bundle locally regardless of --publish.
	bundlePath, writeErr := writeLocalBundle(result.Bundle.EvalBundle)
	if writeErr != nil {
		fmt.Fprintf(ctx.Stderr, "warning: could not write local bundle: %s\n", writeErr)
	}

	if flags.WantsStructuredOutput() {
		command.WriteOutput(ctx.Stdout, flags, map[string]any{
			"bundle":     result.Bundle.EvalBundle,
			"report":     result.Report,
			"conforming": result.Bundle.Conforming,
			"bundlePath": bundlePath,
		})
	} else {
		printReport(ctx, result)
		if bundlePath != "" {
			fmt.Fprintf(ctx.Stdout, "\nBundle written: %s\n", bundlePath)
		}
	}

	if flags.BoolValues["publish"] {
		resp, pubErr := client.SubmitBundle(result.Bundle.EvalBundle)
		if pubErr != nil {
			fmt.Fprintf(ctx.Stderr, "publish failed: %s\n", pubErr)
			return 1
		}
		if !flags.WantsStructuredOutput() {
			tier := resp.TierState
			if tier == "" {
				tier = "self_reported"
			}
			fmt.Fprintf(ctx.Stdout, "Published (tier: %s%s)\n", tier, conformNote(result.Bundle.Conforming))
		}
	}
	return 0
}

func printReport(ctx command.Context, result runner.Result) {
	fmt.Fprintf(ctx.Stdout, "\nResults (%d trials/arm, %s isolation):\n", result.Report.TrialsPerArm, result.Report.Isolation)
	for _, d := range result.Report.Deltas {
		fmt.Fprintf(ctx.Stdout, "  %-24s pass %.0f%%→%.0f%% (%+.0f pp), tokens %+.1f%%\n",
			d.TaskID, d.ControlPassRate*100, d.TreatmentPassRate*100, d.PassRateDelta*100, d.TokenDeltaPct)
	}
	if result.Report.LowNCaveat {
		fmt.Fprintln(ctx.Stdout, "  (low N: <5 trials/arm — treat deltas as directional, not significant)")
	}
}

func conformNote(conforming bool) string {
	if conforming {
		return ""
	}
	return "; not on reference environment → not eligible for Community-reported aggregation"
}

func writeLocalBundle(bundle reg.EvalBundle) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(home, ".skillrank", "bundles")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	name := fmt.Sprintf("%s_%s_%d.json", sanitize(bundle.SkillSlug), sanitize(bundle.SuiteID), time.Now().Unix())
	path := filepath.Join(dir, name)
	buf, err := json.MarshalIndent(bundle, "", "  ")
	if err != nil {
		return "", err
	}
	if err := os.WriteFile(path, buf, 0o644); err != nil {
		return "", err
	}
	return path, nil
}

func detectAgentProvider() string {
	if _, err := exec.LookPath("claude"); err == nil {
		return "claude"
	}
	if _, err := exec.LookPath("codex"); err == nil {
		return "codex"
	}
	return ""
}

func detectAgentVersion(provider string) string {
	cmd := exec.Command(provider, "--version")
	out, err := cmd.Output()
	if err != nil {
		return "unknown"
	}
	fields := strings.Fields(strings.TrimSpace(string(out)))
	for _, f := range fields {
		if len(f) > 0 && (f[0] >= '0' && f[0] <= '9') {
			return f
		}
	}
	return strings.TrimSpace(string(out))
}

func orDash(s string) string {
	if strings.TrimSpace(s) == "" {
		return "(agent default)"
	}
	return s
}

func humanInt(n int) string {
	if n >= 1_000_000 {
		return fmt.Sprintf("%.1fM", float64(n)/1_000_000)
	}
	if n >= 1_000 {
		return fmt.Sprintf("%.0fk", float64(n)/1_000)
	}
	return strconv.Itoa(n)
}

func sanitize(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9', r == '-', r == '_':
			b.WriteRune(r)
		default:
			b.WriteRune('-')
		}
	}
	return b.String()
}

File: crates/skillrank/src/eval.rs
//! `skillrank eval` — run a local forced-mode paired eval on the user's own agent
//! and optionally publish the result bundle.

use crate::commands::new_client;
use crate::flags::Flags;
use skillrank_core as core;
use skillrank_core::runner::{
    self,
    agent::{version_band, CliAgentRunner},
    fixture::{docker_available, GitFixtureProvider, ScriptVerifier},
    Config,
};
use std::process::Command;

pub fn run(args: &[String]) -> i32 {
    let f = Flags::parse(args);
    let Some(reference) = f.positionals.first().cloned() else {
        eprintln!("usage: eval <ref> --suite <id> [--trials N] [--agent claude|codex] [--model M] [--publish]");
        return 2;
    };
    let suite_id = f.value("suite");
    if suite_id.trim().is_empty() {
        eprintln!("error: --suite <id> is required");
        return 2;
    }
    let trials: u32 = f.value("trials").parse().unwrap_or(3).max(1);
    let provider = {
        let p = f.value("agent").trim().to_string();
        if p.is_empty() {
            detect_agent_provider()
        } else {
            p
        }
    };
    if provider != "claude" && provider != "codex" {
        eprintln!("error: could not find a supported agent CLI; install `claude` or `codex`, or pass --agent");
        return 1;
    }
    if !binary_available(&provider) {
        eprintln!("error: agent {provider:?} not found on PATH");
        return 1;
    }

    let client = new_client(&f);

    // Resolve the skill and ensure content for the treatment arm.
    let mut resolved = match client.resolve(&reference) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    if resolved.inline_content.trim().is_empty() && !resolved.raw_content_url.trim().is_empty() {
        match client.fetch_raw_content(&resolved.raw_content_url) {
            Ok(c) => resolved.inline_content = c,
            Err(e) => {
                eprintln!("error: {e}");
                return 1;
            }
        }
    }
    if resolved.inline_content.trim().is_empty() {
        eprintln!("error: registry did not provide skill content to evaluate");
        return 1;
    }

    let suite = match client.get_suite(suite_id) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    let verifiers = match client.fetch_verifiers(suite_id, &suite.version) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("error: could not fetch verifiers for suite {suite_id}: {e}");
            return 1;
        }
    };

    let cfg = Config {
        trials,
        model: f.value("model").to_string(),
    };

    // Cost estimate + confirmation.
    let (est_tokens, est_cost) = runner::estimate_cost(&suite, &cfg);
    if !f.wants_json() {
        println!(
            "Eval plan: skill {} vs no-skill on suite {}@{}",
            resolved.slug, suite.id, suite.version
        );
        println!(
            "  agent: {provider} | model: {} | {trials} trials/arm | {} tasks × 2 arms",
            or_dash(cfg.model.as_str()),
            suite.tasks.len()
        );
        println!(
            "  estimated: ~{} tokens, ~${:.2} on YOUR agent subscription",
            human_int(est_tokens),
            est_cost
        );
        if !docker_available() {
            println!("  note: Docker not detected → worktree isolation; results publish as Self-reported.");
        }
        if !f.bool("yes") && !f.bool("y") && !crate::commands::confirm("Proceed?") {
            println!("Aborted.");
            return 1;
        }
    }

    let agent = CliAgentRunner {
        provider: provider.clone(),
        binary: provider.clone(),
        version: detect_agent_version(&provider),
    };
    let fixtures = GitFixtureProvider::new(&suite.fixture.git_url, &suite.fixture.commit);
    let verifier = ScriptVerifier::new(verifiers);

    let result = match runner::run_eval(&suite, &resolved, &cfg, &agent, &fixtures, &verifier) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    let mut bundle = result.bundle;
    bundle.created_at = core::install::now_rfc3339();

    let bundle_path = write_local_bundle(&bundle);

    if f.wants_json() {
        let out = serde_json::json!({
            "bundle": bundle,
            "report": result.report,
            "conforming": result.conforming,
            "bundlePath": bundle_path.clone().unwrap_or_default(),
        });
        println!("{}", serde_json::to_string_pretty(&out).unwrap());
    } else {
        print_report(&result.report);
        if let Some(p) = &bundle_path {
            println!("\nBundle written: {p}");
        }
    }

    if f.bool("publish") {
        match client.submit_bundle(&bundle) {
            Ok(resp) => {
                if !f.wants_json() {
                    let tier = if resp.tier_state.is_empty() {
                        "self_reported"
                    } else {
                        &resp.tier_state
                    };
                    println!(
                        "Published (tier: {tier}{})",
                        conform_note(result.conforming)
                    );
                }
            }
            Err(e) => {
                eprintln!("publish failed: {e}");
                return 1;
            }
        }
    }
    if !f.wants_json() {
        crate::commands::print_zeroshot_tip();
    }
    0
}

fn print_report(report: &runner::Report) {
    println!(
        "\nResults ({} trials/arm, {} isolation):",
        report.trials_per_arm, report.isolation
    );
    for d in &report.deltas {
        println!(
            "  {:<24} pass {:.0}%→{:.0}% ({:+.0} pp), tokens {:+.1}%",
            d.task_id,
            d.control_pass_rate * 100.0,
            d.treatment_pass_rate * 100.0,
            d.pass_rate_delta * 100.0,
            d.token_delta_pct
        );
    }
    if report.low_n_caveat {
        println!("  (low N: <5 trials/arm — treat deltas as directional, not significant)");
    }
}

fn conform_note(conforming: bool) -> String {
    if conforming {
        String::new()
    } else {
        "; not on reference environment → not eligible for Community-reported aggregation".into()
    }
}

fn write_local_bundle(bundle: &core::EvalBundle) -> Option<String> {
    let home = core::config::home().ok()?;
    let dir = home.join("bundles");
    std::fs::create_dir_all(&dir).ok()?;
    let name = format!(
        "{}_{}_{}.json",
        sanitize(&bundle.skill_slug),
        sanitize(&bundle.suite_id),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0)
    );
    let path = dir.join(name);
    let json = serde_json::to_string_pretty(bundle).ok()?;
    std::fs::write(&path, json).ok()?;
    Some(path.to_string_lossy().to_string())
}

fn detect_agent_provider() -> String {
    for candidate in ["claude", "codex"] {
        if binary_available(candidate) {
            return candidate.to_string();
        }
    }
    String::new()
}

fn binary_available(provider: &str) -> bool {
    Command::new(provider)
        .arg("--version")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

fn detect_agent_version(provider: &str) -> String {
    let Ok(out) = Command::new(provider).arg("--version").output() else {
        return "unknown".to_string();
    };
    let s = String::from_utf8_lossy(&out.stdout);
    for field in s.split_whitespace() {
        if field
            .chars()
            .next()
            .map(|c| c.is_ascii_digit())
            .unwrap_or(false)
        {
            return field.to_string();
        }
    }
    let _ = version_band; // keep import meaningful for downstream
    s.trim().to_string()
}

fn or_dash(s: &str) -> String {
    if s.trim().is_empty() {
        "(agent default)".to_string()
    } else {
        s.to_string()
    }
}

fn human_int(n: i64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.0}k", n as f64 / 1_000.0)
    } else {
        n.to_string()
    }
}

fn sanitize(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' || c == '_' {
                c
            } else {
                '-'
            }
        })
        .collect()
}
